sdk/go/common/resource/plugin/provider.go

```diff
@@ -217,6 +217,34 @@ type DiffResult struct {
 	DeleteBeforeReplace bool                    // if true, this resource must be deleted before recreating it.
 }
 
+// Transforms a `ObjectDiff` into an equivalent `DiffResult`. Because ObjectDiff has no way of
+// determining what should be changed vs replaced, all operations are added to the ChangedKeys set.
+//
+// Any key in ignoreChanged is treated as stable.
+func NewDiffResultFromObjectDiff(obj *resource.ObjectDiff, ignoreChanged ...string) DiffResult {
+	result := DiffResult{}
+	if !obj.AnyChanges() {
+		result.Changes = DiffNone
+		return result
+	}
+
+	ignoreSet := make(map[string]bool, len(ignoreChanged))
+	for _, v := range ignoreChanged {
+		ignoreSet[v] = true
+	}
```

Comment:
I don't think that this is going to do quite what you want. The comment on `Goal.IgnoreChanges` is unfortunately misleading (PR to update it is [here](https://github.com/pulumi/pulumi/pull/8491)): its contents are a list of property paths, not just a list of top-level property keys. [`processIgnoreChanges`](https://github.com/pulumi/pulumi/blob/master/pkg/resource/deploy/step_generator.go#L1164-L1201) has an example of how to work with these paths. I think that the most straightforward way to get this right is to call `processIgnoreChanges` _again_ on `oldInputs` and `newInputs` prior to computing the diff (`processIgnoreChanges` has already been run once using the original inputs from the program; rerunning is necessary b/c `newInputs` is the result of `Check` and may differ from the original inputs due to e.g. default application by the provider). This will also ensure that we don't return `DiffSome` if all changes are ignored.